### ISSUE TITLE
CSV export is not working consistently

### ISSUE THREAD

**Issue by aagarwal-gtr** [metabase/metabase#3748]
We are writing custom SQL parameterized queries and trying to export the data based on the parameter values. Facing the following issue:
1. When we save the query and try to rerun it, the data is retrieved from the db but on trying to export the CSV the error mentioned below occurs.
2. If we make any change to the query and before saving run and export the csv, it works fine.
3. In case of saved parameterized query with default value, every time the csv exported is for the data related to the default value and not for the changed parameter

there is a redirection to empty browser page with:
{"errors":{"query":"field is a required param."}}
 and in the logs:

`Query failure: Assert failed: Unable to substitute 'org_name': param not specified.
Found: null
(contains? params k)
["query_processor$assert_valid_query_result.invokeStatic(query_processor.clj:577)"
"query_processor$assert_valid_query_result.invoke(query_processor.clj:572)"
"query_processor$dataset_query.invokeStatic(query_processor.clj:622)"
"query_processor$dataset_query.invoke(query_processor.clj:579)"
"api.card$run_query_for_card.invokeStatic(card.clj:272)"
"api.card$run_query_for_card.invoke(card.clj:265)"
"api.card$fn__22759$fn__22762.invoke(card.clj:277)"
"api.common.internal$wrap_catch_api_exceptions.invokeStatic(internal.clj:222)"
"api.common.internal$wrap_catch_api_exceptions.invoke(internal.clj:219)"
"api.card$fn__22759.invokeStatic(card.clj:274)"
"api.card$fn__22759.invoke(card.clj:274)"
"middleware$enforce_authentication$fn__30829.invoke(middleware.clj:84)"
"api.routes$fn__30929.invokeStatic(routes.clj:40)"
"api.routes$fn__30929.invoke(routes.clj:40)"
"routes$fn__31914.invokeStatic(routes.clj:24)"
"routes$fn__31914.invoke(routes.clj:24)"
"middleware$log_api_call$fn__30917$fn__30919.invoke(middleware.clj:275)"
"db$_do_with_call_counting.invokeStatic(db.clj:458)"
"db$_do_with_call_counting.invoke(db.clj:452)"
"middleware$log_api_call$fn__30917.invoke(middleware.clj:274)"
"middleware$add_security_headers$fn__30876.invoke(middleware.clj:209)"
"middleware$bind_current_user$fn__30833.invoke(middleware.clj:104)"]
`

- Your browser and the version: (e.x. Chrome 52.1, Firefox 48.0, IE 10)
Chrome Version 54.0.2840.71 (64-bit)
- Your operating system: (e.x. OS X 10, Windows XP, etc)
Ubuntu 15.04
- Your databases: (e.x. MySQL, Postgres, MongoDB, …)
Postgres 
- Metabase version: (e.x. 0.19.3)
 v0.20.3
- Metabase hosting environment: (e.x. Mac app, Elastic Beanstalk, Docker, Heroku, Linux/Ubuntu 12)
Docker
- Metabase internal database: (e.x. H2 (default), MySQL, Postgres)
H2

This is important to us and preventing a major use case. would be great if this is fixed soon :). I am a developer myself so if someone can give me direction I can try fixing it myself.

This is very similar #2799 

Edit:
I get no errors when the queries are ran for display. So a possible solution is to redirect the displayed the data to the csv file and not rerun the query when export button is clicked. Can this be done?

**Comment by milgner** [comment]
Very similar problem here, although the error message is slightly different:

```
Assert failed: Unable to substitute 'credit_note_number': param not specified.
Found: null
(contains? params k)
```

Glad to say that the workaround (change query, download before saving) does work, too.

**Comment by aagarwal-gtr** [comment]
unfortunately an important use case for us is we would like non tech people in our org to go to a question change the filter for exceptional values and download the data as csv. this bug is preventing us from going all out with metabase.

**Comment by salsakran** [comment]
this is a dupe of #3695.

@aagarwal-gtr we're on it.

**Comment by thucnc** [comment]
Hello, I got the same problem 'Assert failed: Unable to substitute 'start_date': param not specified.
Found: null (contains? params k)' when calling csv api as attachment: 
![screen shot 2016-12-21 at 6 28 39 pm](https://cloud.githubusercontent.com/assets/2088055/21387706/70e26b5e-c7ab-11e6-877e-2e3cb2c8fc52.png)
Here is my question on metabase
![screen shot 2016-12-21 at 6 25 48 pm](https://cloud.githubusercontent.com/assets/2088055/21387723/87a814a6-c7ab-11e6-99cb-c374a831af7f.png)

Please help

**Comment by thucnc** [comment]
I can call the api now, with content type as 'application/x-www-form-urlencoded'.

**Comment by aagarwal-gtr** [comment]
@thucnc AFAIK this came in v0.20.1 with #3648 and got fixed in v0.21.0 via #3779 What version are you on?

**Comment by thucnc** [comment]
@aagarwal-gtr I'm using v0.21.0 as you can see in my above attached picture.
